README
Perl Extension "Lingua::Lid"
============================

The Perl extension "Lingua::Lid" provides a Perl interface to Lingua-Systems'
language and character encoding identification library "lid".

The interface is implemented using the XS language and makes the
functionality of the "lid" C library functions available to Perl applications
and modules in a simple to use way.


INSTALLATION

The build script will try to automatically determine the installation base
path of the "lid" C library.

In case you installed "lid" to a custom path, set the environment variable
LID_BASE_DIR accordingly before running Makefile.PL.

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES

 * Perl (>= 5.8.0)
 * lid  (>= 2.0.2)
   Have a look at:
	http://www.lingua-systems.com/language-identifier/lid-library/
 * C build environment (GCC recommended)


COPYRIGHT AND LICENSE

Copyright (C) 2009-2014 Lingua-Systems Software GmbH

This extension is free software. It may be used, redistributed and/or
modified under the terms of the zlib license. For details, see the full text
of the license in the file LICENSE.


Changes
Revision history for the Perl extension Lingua::Lid.

0.04  2014-01-24

    * Makefile.PL: Improved lid library detection
    * lib/Lingua/Lid.pm: Use DynaLoader instead of XSLoader
    * Added support for MS Windows

0.03  2011-01-26

    * t/10_thr-lid_fstr.t, t/11_thr-errstr.t: Updated syntax to be
      compatible with Perl's threads module v1.07 (Perl v5.8.8)

0.02  2010-06-18

    * Added support for lid v3.x.x, prior versions are still supported
    * Added thread support (requires lid v3.x.x)
    * Added thread-safe replacement function Lingua::Lid::errstr() for the
      $Lingua::Lid::errstr package variable
    * Reworked $Lingua::Lid::errstr using Lingua::Lid::Errstr to be
      thread-safe
    * Added threaded tests

0.01  2009-09-30

    * Initial version, developed using lid v2.0.2


# vim: sw=4 sts=4 ai et ft=changelog


Makefile.PL
use 5.008000;
use strict;
use ExtUtils::MakeMaker;

use File::Spec;
use Config;


##
## Check if executed by an automated CPAN Test
##
## In that case, signal the test to stop by not creating a Makefile and
## returning zero. This is necessary because Lingua::Lid implements an
## interface to a commercial library that will most likely not be present
## on any CPAN tester's machine.
##
exit 0 if $ENV{AUTOMATED_TESTING};


##
## Check if the OS is supported by lid
##
unless ( $^O =~ /^(linux|freebsd|solaris|mswin32)$/i )
{
    print <<NOTICE;

lid is currently not available for $^O.

Visit http://www.lingua-systems.com/language-identifier/lid-library/ for a
list of supported operating systems.

NOTICE

    die "OS unsupported\n";
}


##
## Determine lid installation information
##
my %lid = locate_lid();

my %flags = (
    lib => $Config{ccname} eq "cl" ? "" : "-L$lid{lib_dir} -llid",
    obj => $Config{ccname} eq "cl" ? "$lid{lib_dir}\\lid.lib" : "",
    inc => "-I$lid{inc_dir}"
);


my %extra_flags = ();

if ( $^O eq "MSWin32" )
{
    # Workaround for shared library name clashes on Win32
    $extra_flags{DLEXT} = "xs.$Config{dlext}";
}


WriteMakefile(
    NAME          => 'Lingua::Lid',
    VERSION_FROM  => 'lib/Lingua/Lid.pm',
    PREREQ_PM     => {},
    ABSTRACT_FROM => 'lib/Lingua/Lid.pm',
    AUTHOR        => 'Alex Linke <alinke@lingua-systems.com>',
    DEFINE        => '',
    OBJECT        => "Lid$Config{_o} $flags{obj}",
    LIBS          => [ $flags{lib} ],
    INC           => $flags{inc},
    %extra_flags
);


sub locate_lid
{
    my $arch_libdir = $Config{archname} =~ /64(?!int)/ ? "lib64" : "lib";
    my $lib_prefix  = $^O eq "MSWin32" ? "" : "lib";

    my @bases = $^O eq "MSWin32" ?
                    ( 'C:\\lid' )  :
                    ( '/usr', '/usr/local', '/opt/ls' );


    unshift @bases, $ENV{LID_BASE_DIR} if defined $ENV{LID_BASE_DIR};


    foreach my $base (@bases)
    {
        my %i = (
            lib_dir => File::Spec->catfile( $base, $arch_libdir ),
            lib     => File::Spec->catfile(
                        $base, $arch_libdir, "${lib_prefix}lid.$Config{dlext}"
                       ),
            inc_dir => File::Spec->catfile( $base, "include" ),
            inc     => File::Spec->catfile( $base, "include", "lid.h" )
        );

        if ( -f $i{lib} && -f $i{inc} )
        {
            print "lid found beneath $base.\n";
            return %i;
        }
    }


    print STDERR <<NOT_FOUND;

The lid C library could not be found.

Tried: @bases

If you already purchased a copy of lid and decided to install it to a non
default location, set the LID_BASE_DIR environment variable to point to
your installation's base directory and rerun Makefile.PL:

    export LID_BASE_DIR=/path/           (Unix, sh)
    setenv LID_BASE_DIR /path/           (Unix, csh)
    set    LID_BASE_DIR=C:\\path          (Win32)

    perl Makefile.PL

Otherwise, point your browser to lid's website for further information:

    http://www.lingua-systems.com/language-identifier/lid-library/

NOT_FOUND

    exit(1);
}


# vim: sts=4 sw=4 ts=4 ai et


lib/Lingua/Lid/Errstr.pm
package Lingua::Lid::Errstr;

#
# Copyright (c) 2010-2014 by Lingua-Systems Software GmbH
#

use strict;
use Tie::Scalar;
use Carp qw/confess/;

use Lingua::Lid;

our @ISA = qw/Tie::Scalar/;

our $VERSION = '0.01';


sub TIESCALAR
{
    my $self;
    return bless \$self, __PACKAGE__;
}


sub FETCH
{
    return Lingua::Lid::errstr();
}


sub STORE
{
    confess('Storing values to $Lingua::Lid::errstr prohibited');
}


1;


=head1 NAME

Lingua::Lid::Errstr - provides $Lingua::Lid::errstr

=head1 SYNOPSIS

    use Lingua::Lid;

    print $Lingua::Lid::errstr, "\n";

    print Lingua::Lid::errstr(), "\n";  # preferred

=head1 DESCRIPTION

B<Lingua::Lid::Errstr> implements a thread-safe replacement of the former
package variable B<$Lingua::Lid::errstr>.

=head1 NOTES

B<$Lingua::Lid::errstr> is provided for backwards compatibility with
B<Lingua::Lid> v0.01. It may be removed in a future release. Any new code
using B<Lingua::Lid> should utilize the function B<Lingua::Lid::errstr>().

=head1 AUTHOR

Alex Linke E<lt>alinke@lingua-systems.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010-2014 Lingua-Systems Software GmbH

This extension is free software. It may be used, redistributed and/or
modified under the terms of the zlib license. For details, see the full text
of the license in the file LICENSE.

=cut


# vim: sts=4 sw=4 ts=4 et ai


lib/Lingua/Lid.pm
package Lingua::Lid;

#
# Copyright (C) 2009-2014 Lingua-Systems Software GmbH
#

use 5.008000;
use strict;

require Exporter;
require DynaLoader;

our @ISA = qw/DynaLoader Exporter/;

use Lingua::Lid::Errstr;


our $VERSION = '0.04';


our @EXPORT      = ();
our %EXPORT_TAGS = ( 'all' => [ qw/lid_version lid_version_ct
                                   lid_ffile lid_fstr/ ] );
our @EXPORT_OK   = ( @{ $EXPORT_TAGS{'all'} } );


# ...for backward compatibility
tie our $errstr, 'Lingua::Lid::Errstr';


# Workaround for shared library name clashes on Win32
local $DynaLoader::dl_dlext = "xs.$DynaLoader::dl_dlext" if $^O eq "MSWin32";


bootstrap Lingua::Lid;


1;


=head1 NAME

Lingua::Lid - Interface to the language and encoding identifier "lid"


=head1 SYNOPSIS

    use Lingua::Lid qw/:all/;

    # Identify the language and character encoding of...

    # ...a string
    $result = lid_fstr("This is a short English sentence.");

    # ...a plain text file
    $result = lid_ffile("/path/to/a/file.txt");

    # ...if $result is undef, an error occurred:
    die Lingua::Lid::errstr() unless $result;

    print "Lingua::Lid v$Lingua::Lid::VERSION, using lid v",
        lid_version(), "\n";


=head1 DESCRIPTION

The Perl extension B<Lingua::Lid> provides a Perl interface to Lingua-Systems'
language and character encoding identification library B<lid>, which is
required to build and use this extension.

The interface is implemented using the XS language and makes the functionality
of the B<lid> C library functions available to Perl applications and modules
in a simple to use way.

B<Lingua::Lid> is thread-safe an can be used my more than one thread
simultaneously, if compiled with B<lid> v3.0.0 or above.

This man page covers the usage of the B<Lingua::Lid> Perl extension only - for
more information on B<lid> and a list on supported languages and character
encodings, have a look at its manual, which is both included in its
distribution and freely available under
L<http://www.lingua-systems.com/language-identifier/lid-library/>.

B<Lingua::Lid> aims to stick with the C interface as close as reasonable - but
with respect to common Perl conventions. Have a look at "COMPARISON TO THE C
INTERFACE" for details.

=head1 EXPORTS

No symbols are exported by default.

Any function needed must either be requested for import explicitly or the
export tag C<:all> may be used to import symbols for all provided functions:

  use Lingua::Lid qw/lid_ffile lid_fstr/; # or
  use Lingua::Lid qw/:all/;

The function Lingua::Lid::errstr() is not exportable and has to be called
with its full package name.

=head1 FUNCTIONS

=head2 lid_fstr( C<$string> )

Mnemonic: "Language and encoding identification... from string"

This function takes a C<$string> as an argument and identifies its language
and encoding.
It returns a hash reference containing the results. See
S<IDENTIFICATION RESULTS DATA STRUCTURE> for details.

If an error occurs, the function returns C<undef>. Use Lingua::Lid::errstr()
to obtain an appropriate message describing the error.

=head2 lid_ffile( C<$file> )

Mnemonic: "Language and encoding identification... from file"

This function takes a plain text C<$file>'s path as an argument and identifies
its language and encoding.  It returns a hash reference containing the
results. See S<IDENTIFICATION RESULTS DATA STRUCTURE> for details.

If an error occurs, the function returns C<undef>. Use Lingua::Lid::errstr()
to obtain an appropriate message describing the error.

=head2 lid_version( )

This function returns the version of the B<lid> C library that is currently
loaded (I<runtime> version).

=head2 lid_version_ct( )

This function returns the version of the B<lid> C library that B<Lingua::Lid>
has been compiled with (I<compile time> version).

=head1 IDENTIFICATION RESULTS DATA STRUCTURE

The functions lid_fstr() and lid_ffile() return a I<hash reference>
containing the results of the language and encoding identification.

The hash reference contains the following keys:

=over 4

=item I<language>

The language's name (in English), i.e. "German", "French", "English".

=item I<isocode>

The language's ISO 639-3 code, i.e. "deu", "fra", "eng".

=item I<encoding>

The character encoding, i.e. "UTF-8", "ISO-8859-1", "UTF-32BE".

=back

  $result = {
                'language'  =>  'English',
                'isocode'   =>  'eng',
                'encoding'  =>  'ASCII'
            };


=head1 ERROR HANDLING

The functions lid_fstr() and lid_ffile() return C<undef> if an error occurs.
Lingua::Lid::errstr() can be used to obtain an appropriate message describing
the last occurred error.

Have a look at B<lid>'s manual for a list of all error messages.

=over 4

=item NOTE:

The C<$Lingua::Lid::errstr> variable is still supported and thread-safe, too.
Internally it is tied to Lingua::Lid::errstr() using
B<Lingua::Lid::Errstr>.
However, as of B<Lingua::Lid> v0.02 Lingua::Lid::errstr() is preferred and
should be used in any new code. C<$Lingua::Lid::errstr> may be removed in a
future release.

=back

=head1 COMPARISON TO THE C INTERFACE

B<Lingua::Lid>'s function lid_fstr() and lid_ffile() behave exactly as their
B<lid> counterparts in C.

The C functions lid_fnstr() and lid_fwstr() are not needed, use the
B<Lingua::Lid> function lid_fstr() in any Perl code instead.

The C function lid_strerror() and the per-thread pseudo-variable C<lid_errno>
are not needed. Rather than returning a pointer to C<NULL>, B<Lingua::Lid>'s
lid_fstr() and lid_ffile() return C<undef> on errors.
Lingua::Lid::errstr() can be used to obtain an appropriate message describing
the last occurred error.

B<lid>'s function lid_version_string() is available as lid_version() in
B<Lingua::Lid>.

The C defines C<LID_VERSION_STRING> (C<LID_VERSION> in lid v2.x.x) is not
available in B<Lingua::Lid>, use lid_version_ct() instead.

B<Lingua::Lid>'s results data structure sticks to the C C<lid_t *> structure
as close as possible. See "IDENTIFICATION RESULTS DATA STRUCTURE" above.

=head1 EXAMPLES

  use strict;
  use Lingua::Lid qw/lid_fstr lid_version/;

  print "Lingua::Lid v$Lingua::Lid::VERSION, using lid v",
    lid_version(), "\n";

  my @strings =
  (
      "This is a short English sentence.",
      "Dies ist ein kurzer deutscher Satz.",
      " "
  );

  foreach my $string (@strings)
  {
      if (my $r = lid_fstr($string))
      {
          print join(" - ", $r->{language}, $r->{isocode},
                            $r->{encoding}), "\n";
      }
      else
      {
          print "lid_fstr() failed: ", Lingua::Lid::errstr(), "\n";
      }
  }

The program above produces the following output:

  Lingua::Lid v0.02, using lid v3.0.0
  English - eng - ASCII
  German - deu - ASCII
  lid_fstr() failed: Insufficient input length

=head1 BUGS

None known.

Please report bugs either using CPAN's bug tracker or to
E<lt>perl@lingua-systems.comE<gt>.


=head1 SEE ALSO

=over 4

=item *

Lingua::Lid's website:
L<http://www.lingua-systems.com/language-identifier/Lingua-Lid-Perl-extension/>

=item *

lid's website:
L<http://www.lingua-systems.com/language-identifier/lid-library/>

=item *

B<lid>'s manual (available in English and German)

=back


=head1 AUTHOR

Alex Linke E<lt>alinke@lingua-systems.comE<gt>


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009-2014 Lingua-Systems Software GmbH

This extension is free software. It may be used, redistributed and/or
modified under the terms of the zlib license. For details, see the full text
of the license in the file LICENSE.

=cut


# vim: sts=4 sw=4 ts=4 ai et tw=78
